crates/sqllib/src/string_interner.rs

```diff
@@ -0,0 +1,616 @@
+
+    /// The first input stream contains strings to be interned.
+    /// The second input stream contains queries for interned strings. It is joined with the first stream
+    /// to produce an output stream of un-interned strings.
+    pub fn interner_test_circuit(
+        path: &Path,
+        checkpoint: Option<Uuid>,
+    ) -> (
+        DBSPHandle,
+        (
+            ZSetHandle<SqlString>,
+            ZSetHandle<SqlString>,
+            OutputHandle<OrdZSet<SqlString>>,
+        ),
+    ) {
+        let (circuit, handles) = Runtime::init_circuit(
+            CircuitConfig::with_workers(8).with_storage(
+                CircuitStorageConfig::for_config(
+                    StorageConfig {
+                        path: path.display().to_string(),
+                        cache: Default::default(),
+                    },
+                    StorageOptions::default(),
+                )
+                .unwrap()
+                .with_init_checkpoint(checkpoint),
+            ),
+            move |circuit| {
+                let (input_strings, hinput_strings) = circuit.add_input_zset::<SqlString>();
+                input_strings.set_persistent_mir_id(&"input_strings".to_string());
+
+                let (queries, hqueries) = circuit.add_input_zset::<SqlString>();
+                queries.set_persistent_mir_id(&"queries".to_string());
+
+                build_string_interner(input_strings.map(|s| Tup1(s.clone())));
+
+                let output_strings = input_strings
+                    .map_index(|s| (s.clone(), intern_string(s)))
+                    .join(
+                        &queries.map_index(|q| (q.clone(), ())),
+                        |_, intern_string_id, _| unintern_string(intern_string_id).unwrap(),
+                    );
+
+                Ok((hinput_strings, hqueries, output_strings.output()))
+            },
+        )
+        .unwrap();
+        (circuit, handles)
+    }
+
+    /// Push queries to the circuit, force a step, and check that the output matches the queries.
+    fn query<'a, I>(
+        circuit: &mut DBSPHandle,
+        hqueries: &ZSetHandle<SqlString>,
+        houtput_strings: &OutputHandle<OrdZSet<SqlString>>,
+        queries: I,
+    ) where
+        I: IntoIterator<Item = &'a str>,
+    {
+        let mut queries = queries.into_iter().map(SqlString::from).collect::<Vec<_>>();
+
+        let mut tuples = queries
+            .iter()
+            .map(|s| Tup2(s.clone(), 1))
+            .collect::<Vec<_>>();
+        hqueries.append(&mut tuples);
+
+        circuit.step().unwrap();
+        let output = houtput_strings.consolidate();
+        let mut output = output.iter().map(|(s, _, _)| s.clone()).collect::<Vec<_>>();
+        output.sort();
+
+        queries.sort();
+        assert_eq!(output, queries);
+
+        let mut tuples = queries
+            .iter()
+            .map(|s| Tup2(s.clone(), -1))
+            .collect::<Vec<_>>();
+        hqueries.append(&mut tuples);
```

Comment:
what are these used for?

Reply:
It's explained in a comment somewhere, it's used to ask the circuit to unintern specific strings as part of a test.